content/en/docs/zero-code/obi/context-propagation.md

```diff
@@ -0,0 +1,141 @@
+When an incoming request is handled, any outgoing requests, typically done in async fashion to
+avoid holding up the event loop, are added on the event queue loop as separate requests. To correctly
+propagate context between incoming and outgoing requests, we must know which async operation was 
+scheduled by what other async operation.
+
+Prior to `Node.js` 20 we used to track the async event loop correlation, by injecting probes into
+`EmitAsyncInit` and `AsyncReset`. However, from `Node.js` 20 and onward, more of the runtime 
+handling was rewritten in JavaScript, and the eBPF probe approach misses number of critical correlation
+steps.
+
+To correctly track `Node.js`, OBI injects a small agent that hooks into `serverEmit`, `socketConnect` and
+`socketWrite`. This agent is automatically injected in running processes, through the 
+`Node.js` debugger interface, and it has 0 dependencies. It extracts the incoming and outgoing
+file descriptors for the requests, and it communicates them to the eBPF side by performing a
+fake file operation on `/dev/null`.
+
+The eBPF side tracks the fake file access and establishes an internal in-memory map of 
+incoming to outgoing file descriptor mappings. When an outgoing request happens, OBI looks up
+the `Node.js` map to find the parent file descriptor, which is then matched to an incoming request.
+
+`Node.js` correlation doesn't for for `HTTP2`/`gRPC`.
+
+### nginx
+
+`nginx` has a unique threading model, where number of incoming requests and handled by a 
+custom thread pool of `upstream` handlers. To support `nginx` context propagation, OBI injects
+two probes, a probe in `ngx_http_upstream_init` and a return probe on `ngx_event_connect_peer`.
+
+At the time `ngx_http_upstream_init` is invoked, we start tracking the connection information
+of the incoming request, while when `ngx_event_connect_peer` returns, we create a mapping
+of the incoming request to the outgoing file descriptor. At the time OBI handles the outgoing 
+request, it looks up the outgoing call file descriptor and fetches the connection information
+of the incoming request, managing to correlate both. In a sense, this is very similar to how
+we lookup the information for the `Node.js` event loop.
+
+`nginx` correlation doesn't for for `HTTP2`/`gRPC`.
+
+### Generic approach
+
+The fallback approach of incoming to outgoing request correlation, for the purpose of
+distributed trace context propagation, relies on thread ID correlation. The following
+section outlines all the techniques we currently employ to match the requests by
+thread ID:
+
+1. Same thread. OBI tries to detect if the incoming and outgoing requests are handled
+by the same thread. It detects if multiple current outgoing requests are handled by
+the same thread and marks the correlation information as invalid. This helps us
+prevent incorrect correlation, when the application framework handles multiple 
+connections on the same thread.
```

Comment:
👍 that was my biggest question for asyncio, if OBI would generate incorrect parent-child relationships